File: benchmarks/matmul/java/Matmul.java
public class Matmul {
    public static void main(String[] args) {
        int n = args.length > 0 ? Integer.parseInt(args[0]) : 400;

        double[][] a = new double[n][n];
        double[][] b = new double[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                a[i][j] = (i * 3 + j) % 13;
                b[i][j] = (i + j * 2) % 17;
            }
        }

        double[][] c = new double[n][n];
        for (int i = 0; i < n; i++) {
            for (int k = 0; k < n; k++) {
                double aik = a[i][k];
                for (int j = 0; j < n; j++) {
                    c[i][j] += aik * b[k][j];
                }
            }
        }

        double expected = 0.0;
        for (int k = 0; k < n; k++) expected += a[0][k] * b[k][0];
        if (c[0][0] != expected) {
            System.err.println("self-check failed: c[0][0] mismatch");
            System.exit(1);
        }

        System.out.printf("%.2f%n", c[n - 1][n - 1]);
    }
}

File: benchmarks/matmul/rust/matmul.rs
use std::env;

fn main() {
    let n: usize = env::args().nth(1).map(|s| s.parse().unwrap()).unwrap_or(400);

    let mut a = vec![0.0f64; n * n];
    let mut b = vec![0.0f64; n * n];
    for i in 0..n {
        for j in 0..n {
            a[i * n + j] = ((i * 3 + j) % 13) as f64;
            b[i * n + j] = ((i + j * 2) % 17) as f64;
        }
    }

    let mut c = vec![0.0f64; n * n];
    for i in 0..n {
        for k in 0..n {
            let aik = a[i * n + k];
            for j in 0..n {
                c[i * n + j] += aik * b[k * n + j];
            }
        }
    }

    let expected: f64 = (0..n).map(|k| a[k] * b[k * n]).sum();
    assert_eq!(c[0], expected, "self-check failed: c[0][0] mismatch");

    println!("{:.2}", c[(n - 1) * n + (n - 1)]);
}
